Flickering on Raspberry Pi

### Dupe Check

- [X] I have [searched Warp bugs](https://github.com/warpdotdev/warp/issues?q=is%3Aissue+label%3ABugs) and there are no duplicates

### Describe the bug

So essentially when I move the mouse around it often flickers and window becomes transparent.

### To reproduce

1. Install Warp on Ubuntu 24.04 on a Raspberry Pi
2. Log in
3. Move the mouse around and see bug

### Expected behavior

I would expect to warp work normally.

### Screenshots

_No response_

### Operating system

Linux

### Operating system and version

Ubuntu 24.04

### Shell Version

bash 5.2.21

### Current Warp version

v0.2024.04.30.08.02.stable_02

### Regression

No, this bug or issue has existed throughout my experience using Warp

### Recent working Warp date

_No response_

### Additional context

I use Wayland on GNOME. I did see that this version fixes flickering on Intel UHD 620. maybe something similar would be needed for Raspberry Pi.




### Does this block you from using Warp daily?

Yes, this issue prevents me from using Warp daily.

### Is this a Warp specific issue? (i.e. does it happen in Terminal, iTerm, Kitty, etc.)

Yes, this I confirmed this only happens in Warp, not other terminals.

### Warp Internal (ignore): linear-label:b9d78064-c89e-4973-b153-5178a31ee54e

None

The problem is still there unfortunately. Trying to run "WGPU_BACKEND=gl warp-terminal" from the linux terminal results in this:

> 17:21:57 [INFO] Spawning terminal server process...
> 17:21:57 [INFO] Running terminal server...
> Error: Address already in use (os error 98)
> 17:21:57 [INFO] Running app with windowing system: Wayland
> 17:21:57 [ERROR] Failed to acquire default Secret Service collection: unknown error
> 17:21:57 [INFO] Initializing crash reporting Some("linux_stable_release") with tag "v0.2025.01.08.08.02.stable_04"...
> 17:21:57 [INFO] Starting warp with channel state ChannelState { channel: Stable, app_id: AppId { qualifier: "dev", organization: "warp", application_name: "Warp" }, additional_features: {}, firebase_api_key: "AIzaSyBdy3O3S9hrdayLJxJ7mriBR4qgUaUygAs", server_root_url: "https://app.warp.dev", ws_server_url: "wss://rtc.app.warp.dev/graphql/v2", session_sharing_server_url: Some("wss://sessions.app.warp.dev"), rudderstack_write_key: "2iyG8S1RaV4XTX6SEypetFSEbLK", rudderstack_root_url: "https://warpianwzlfqdq.dataplane.rudderstack.com", releases_base_url: "https://releases.warp.dev", sentry_url: "https://0195a81da0714f55a93ee4624825f9ec@o540343.ingest.sentry.io/5658526", logfile_name: "warp.log", show_autoupdate_menu_items: true } and version Some("v0.2025.01.08.08.02.stable_04")
> 17:21:57 [INFO] Enabling use of lazy scene building
> 17:21:57 [INFO] Start to send telemetry events to RudderStack
> 17:21:57 [INFO] Performance metrics collector started
> 17:21:57 [INFO] Dropping empty RudderStack telemetry batch
> 17:21:57 [INFO] Flushed telemetry events.
> 17:21:57 [INFO] Initializing app services
> 17:21:57 [INFO] Start to send telemetry events to RudderStack
> 17:21:57 [INFO] Successfully flushed events to rudder from disk
> 17:21:58 [WARN] 	EGL context: -robust access
> 17:21:58 [WARN] SQLite error 283 (A WAL mode database file was recovered): recovered 309 frames from WAL file /home/tomask/.local/state/warp-terminal/warp.sqlite-wal
> 17:21:58 [INFO] Connecting to SQLite database
> 17:21:58 [INFO] Checking for update on channel stable_release. Update id is FqyTxPT
> 17:21:58 [INFO] dispatching global action for root_view:open_from_restored
> 17:21:58 [INFO] Fetching channel versions (without changelogs) from Warp server
> 17:21:58 [INFO] Computing available system fonts
> 17:21:58 [WARN] Could not find current AIAgentApiType value in dropdown option list
> 17:21:58 [INFO] Computing available system fonts
> 17:21:58 [WARN] Could not find current AIAgentApiType value in dropdown option list
> 17:21:58 [INFO] Start to send telemetry events to RudderStack
> 17:21:58 [INFO] Block finished with new state DoneWithNoExecution
> 17:21:58 [INFO] Block finished with new state DoneWithNoExecution
> 17:21:58 [INFO] Block finished with new state DoneWithNoExecution
> 17:21:58 [INFO] Block finished with new state DoneWithNoExecution
> 17:21:58 [INFO] Block finished with new state DoneWithNoExecution
> 17:21:58 [INFO] Fetching channel versions and changelogs from Warp server
> 17:21:58 [INFO] dispatching typed action: HandleFocusChange
> 17:21:58 [WARN] Action HandleFocusChange was dispatched, but no view handled it
> 17:21:58 [INFO] dispatching typed action: HandleFocusChange
> 17:21:58 [WARN] Action HandleFocusChange was dispatched, but no view handled it
> 17:21:58 [WARN] Tried to update block filter query without active_filter_editor_block_index set
> 17:21:58 [INFO] Loaded 34 font families
> 17:21:58 [INFO] Loaded 34 font families
> 17:21:58 [INFO] fetching updated cloud objects
> 17:21:58 [INFO] Start to send telemetry events to RudderStack
> 17:21:58 [INFO] dispatching typed action: HandleFocusChange
> 17:21:58 [WARN] Action HandleFocusChange was dispatched, but no view handled it
> 17:21:58 [INFO] fetching updated cloud objects
> 17:21:58 [INFO] dispatching typed action: HandleFocusChange
> 17:21:58 [WARN] Action HandleFocusChange was dispatched, but no view handled it
> 17:21:58 [INFO] dispatching typed action: HandleFocusChange
> 17:21:58 [WARN] Action HandleFocusChange was dispatched, but no view handled it
> 17:21:58 [INFO] dispatching typed action: HandleFocusChange
> 17:21:58 [WARN] Action HandleFocusChange was dispatched, but no view handled it
> 17:21:58 [INFO] dispatching typed action: HandleFocusChange
> 17:21:58 [WARN] Action HandleFocusChange was dispatched, but no view handled it
> 17:21:58 [INFO] dispatching typed action: HandleFocusChange
> 17:21:58 [WARN] Action HandleFocusChange was dispatched, but no view handled it
> 17:21:58 [INFO] Starting shell /bin/bash
> 17:21:58 [INFO] Successfully spawned child bash process with pid 35677
> 17:21:58 [INFO] Successfully spawned tty with pid: 35677
> 17:21:58 [INFO] Reconstructing database, but it already exists
> 17:21:58 [INFO] Connecting to SQLite database
> 17:21:58 [INFO] SQLite Writer is resumed
> 17:21:58 [INFO] Received InitShell hook
> 17:21:58 [ERROR] Failed to open window: Failed to create surface for any enabled backend: {}
> 17:21:58 [INFO] WindowId(0) will close
> 17:21:58 [INFO] dispatching global action for workspace:save_app
> 17:21:58 [INFO] application will terminate
> 17:21:58 [INFO] Shutting down SQLite writer thread
> 17:21:58 [INFO] Shut down SQLite writer in 16.717673ms
> 17:21:58 [INFO] Writing queued events to disk because telemetry is enabled.
> 17:21:58 [INFO] Successfully wrote telemetry events to disk
> 17:21:58 [INFO] Tearing down terminal server...
> 17:21:58 [INFO] Received empty message; assuming the connection has been closed.
> 17:21:58 [INFO] Tearing down app services...
> 17:21:58 [INFO] Uninitializing crash reporting...
> 17:21:58 [INFO] Received channel versions from Warp server: dev: ChannelVersion { version_info: VersionInfo { version: "v0.2025.01.15.08.02.dev_00", update_by: None, soft_cutoff: Some("v0.2023.05.12.08.03.dev_00") }, overrides: [] }; preview: ChannelVersion { version_info: VersionInfo { version: "v0.2025.01.15.08.02.preview_00", update_by: None, soft_cutoff: None }, overrides: [] }; canary: ChannelVersion { version_info: VersionInfo { version: "v0.2022.09.29.08.08.canary_00", update_by: None, soft_cutoff: None }, overrides: [] }; beta: ChannelVersion { version_info: VersionInfo { version: "v0.2024.12.18.08.02.beta_00", update_by: None, soft_cutoff: None }, overrides: [] }; stable: ChannelVersion { version_info: VersionInfo { version: "v0.2025.01.08.08.02.stable_04", update_by: None, soft_cutoff: Some("v0.2024.11.25.16.32.stable_00") }, overrides: [VersionOverride { predicate: TargetOS(Windows), version_info: VersionInfo { version: "v0.2024.12.18.08.02.stable_05", update_by: None, soft_cutoff: None } }] }

Any idea what to do? Running Raspberry Pi 5 4GB with Wayland and latest updates.

--- a/crates/warpui/src/rendering/wgpu/resources.rs
+++ b/crates/warpui/src/rendering/wgpu/resources.rs
@@ -535,6 +535,20 @@ pub fn adapter_has_rendering_offset_bug(adapter_info: &wgpu::AdapterInfo) -> boo
         .any(|model| adapter_info.name.contains(model))
 }
 
+/// Returns whether the adapter is the Broadcom V3D Vulkan driver (V3DV).
+///
+/// The V3D Vulkan driver on Raspberry Pi and similar ARM SBCs have panics and flickering issues.
+/// The logs have warnings about these drivers missing the `FULL_DRAW_INDEX_UINT32` downlevel flag
+/// but if it's unclear if that is the actual cause. See:
+/// https://github.com/warpdotdev/warp/issues/10618
+/// https://github.com/warpdotdev/warp/issues/4879
+fn is_v3d_vulkan_adapter(adapter_info: &wgpu::AdapterInfo) -> bool {
+    cfg!(target_os = "linux")
+        && adapter_info.backend == wgpu::Backend::Vulkan
+        && adapter_info.driver.contains("V3DV")
+        && adapter_info.name.starts_with("V3D 4.2.14.")
+}
+
 /// Checks whether the provided adapter info describes a lavapipe
 /// (Vulkan llvmpipe) adapter that may not work properly with warpui.
 fn is_older_lavapipe_adapter(adapter_info: &wgpu::AdapterInfo) -> bool {
@@ -712,6 +726,11 @@ fn adapter_stability_sort_func(
         && !is_vulkan_nvidia_adapter(&adapter_info)
     {
         log::info!("Deprioritizing non-NVIDIA Vulkan adapter (the PRIME performance profile is likely enabled)");
+        return AdapterSupport::Unsupported;
+    }
+
+    if is_v3d_vulkan_adapter(&adapter_info) {
+        log::warn!("Deprioritizing Vulkan-backed V3D adapter");
         return AdapterSupport::Unsupported;
     }
 

